akka-persistence/src/main/scala/akka/persistence/journal/PersistencePluginProxy.scala

```diff
@@ -0,0 +1,186 @@
+          Persistence(context.system).journalFor(targetPluginId)
+        case SnapshotStore ⇒
+          log.info("Starting target snapshot-store [{}]", targetPluginId)
+          Persistence(context.system).snapshotStoreFor(targetPluginId)
+      }
+      context.become(active(target, targetAtThisNode = true))
+    } else {
+      val targetAddressKey = s"target-${pluginType.qualifier}-address"
+      val targetAddress = config.getString(targetAddressKey)
+      if (targetAddress != "") {
+        try {
+          log.info("Setting target {} address to {}", pluginType.qualifier, targetAddress)
+          PersistencePluginProxy.setTargetLocation(context.system, AddressFromURIString(targetAddress))
+        } catch {
+          case _: URISyntaxException ⇒ log.warning("Invalid URL provided for target {} address: {}", pluginType.qualifier, targetAddress)
+        }
+      }
+
+      context.system.scheduler.scheduleOnce(initTimeout, self, InitTimeout)(context.dispatcher)
+    }
+  }
+
+  private val selfAddress: Address =
+    context.system.asInstanceOf[ExtendedActorSystem].provider.getDefaultAddress
+
+  private def timeoutException() = new TimeoutException(s"Target ${pluginType.qualifier} not initialized. " +
+    s"Use `PersistencePluginProxy.setTargetLocation` or set `target-${pluginType.qualifier}-address`")
+
+  def receive = init
+
+  def init: Receive = {
+    case TargetLocation(address) ⇒
+      context.setReceiveTimeout(1.second) // for retries
+      context.become(identifying(address))
+    case InitTimeout ⇒
+      log.info("Initialization timed-out (after {}), Use `PersistencePluginProxy.setTargetLocation` or set `target-{}-address`", initTimeout, pluginType.qualifier)
+      context.become(initTimedOut)
+      unstashAll() // will trigger appropriate failures
+    case msg ⇒
+      stash()
+  }
+
+  def becomeIdentifying(address: Address): Unit = {
+    sendIdentify(address)
+    context.setReceiveTimeout(1.second) // for retries
+    context.become(identifying(address))
+  }
+
+  def sendIdentify(address: Address): Unit = {
+    val sel = context.actorSelection(RootActorPath(address) / "system" / targetPluginId)
+    log.info("Trying to identify target {} at {}", pluginType.qualifier, sel)
+    sel ! Identify(targetPluginId)
+  }
+
+  def identifying(address: Address): Receive = ({
+    case ActorIdentity(`targetPluginId`, Some(target)) ⇒
+      log.info("Found target {} at [{}]", pluginType.qualifier, address)
+      context.setReceiveTimeout(Duration.Undefined)
+      context.watch(target)
+      unstashAll()
+      context.become(active(target, address == selfAddress))
+    case _: ActorIdentity ⇒ // will retry after ReceiveTimeout
+    case ReceiveTimeout ⇒
+      sendIdentify(address)
+  }: Receive).orElse(init)
+
+  def active(targetJournal: ActorRef, targetAtThisNode: Boolean): Receive = {
+    case TargetLocation(address) ⇒
+      if (targetAtThisNode && address != selfAddress)
+        becomeIdentifying(address)
```

Comment:
to be on the safe side I think you need to handle `Terminated` also in the other Receive states, in case we don't get a Terminated from an "old" journal. In the other places it can just be ignored (but it must be handled).